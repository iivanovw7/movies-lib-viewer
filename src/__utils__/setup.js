/**
 * Module contains tests setup
 * @module __utils__/setup
 * @author Igor Ivanov
 */
import 'core-js/stable';
import 'regenerator-runtime/runtime';
import Enzyme, { configure } from 'enzyme';
import jsdom from 'jsdom';

// eslint-disable-next-line import/no-extraneous-dependencies
const Adapter = require('enzyme-adapter-react-16');

Enzyme.configure({ adapter: new Adapter() });

const dom = new jsdom.JSDOM('<!DOCTYPE html><html><body></body></html>', {
  url: 'http://localhost:4428',
  pretendToBeVisual: true,
});
const win = dom.window;

global.document = win.document;
global.window = win;
global.localStorage = win.localStorage;
global.requestAnimationFrame = win.requestAnimationFrame;
global.cancelAnimationFrame = win.cancelAnimationFrame;

// eslint-disable-next-line no-shadow
global.window.IntersectionObserver = class IntersectionObserver {
  // eslint-disable-next-line no-useless-constructor,require-jsdoc,no-empty-function
  constructor() {}

  // eslint-disable-next-line class-methods-use-this,require-jsdoc
  observe(cb) {
    return cb;
  }

  // eslint-disable-next-line class-methods-use-this,require-jsdoc
  unobserve() {
    return null;
  }
};

global.window.IntersectionObserverEntry = {};
global.window.IntersectionObserverEntry.prototype = {
  intersectionRatio: {},
};

configure({ adapter: new Adapter() });
