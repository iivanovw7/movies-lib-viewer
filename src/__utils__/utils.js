/**
 * Module contains test utils
 * @module __utils__/utils
 * @author Igor Ivanov
 */
import chalk from 'chalk';
/* eslint-disable */
export const testName = (title, description) => `[${chalk.yellow(title)}]: ${description}`;
/* eslint-enable */
