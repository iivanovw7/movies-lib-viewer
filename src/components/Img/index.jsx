/**
 * Module renders an image with parameters
 * @module components/Img
 * @author Igor Ivanov
 */
import * as PropTypes from 'prop-types';
import React from 'react';
import styled from 'styled-components';

export const StyledImg = styled.img`
  ${(props) => props.styling};
`;

/**
 * Creates image component.
 * @method
 * @param {Object} props
 *  contains component props
 *  @see {@link module:components/Img~Img.propTypes}
 * @return {Node} React component with children.
 * @constructor
 */
const Img = ({ src, alt, id, styling, onClick, handleLoad, className }) => (
  <StyledImg
    src={src}
    alt={alt}
    data-id={id}
    onLoad={handleLoad}
    onClick={onClick}
    styling={styling}
    className={className}
  />
);

/**
 * @name Img.propTypes
 * @type {Object}
 * @param {Object} props - React PropTypes
 * @property {string | Object} props.src - image src.
 * @property {string} props.alt - alt string.
 * @property {string} props.id - id string.
 * @property {string} [props.className = ''] - className string.
 * @property {Array.<string>} [props.styling = []] - image additional styles.
 * @return {Array} React propTypes
 */
Img.propTypes = {
  src: PropTypes.oneOfType([PropTypes.string, PropTypes.object]).isRequired,
  alt: PropTypes.string.isRequired,
  styling: PropTypes.array,
  className: PropTypes.string,
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  handleLoad: PropTypes.func,
  onClick: PropTypes.func,
};

Img.defaultProps = {
  styling: [],
  className: '',
};

export default Img;
