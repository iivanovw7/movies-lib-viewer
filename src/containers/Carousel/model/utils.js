/**
 * Module contains carousel page utils
 * @module containers/Carousel/model/utils
 */

import setting from '../../../config/data';

/**
 * Calculates previous position.
 * @param {number} length - slides length.
 * @param {number} current - current slide.
 * @param {number} [skip = 0] - slide to be skipped.
 * @return {number} previous slide position.
 */
export function previous(length, current, skip = 0) {
  return Number((current - 1 - skip + length) % length);
}

/**
 * Calculates next position.
 * @param {number} length - slides length.
 * @param {number} current - current slide.
 * @param {number} [skip = 0] - slide to be skipped.
 * @return {number} next slide position.
 */
export function next(length, current, skip = 0) {
  return Number((current + 1 + skip) % length);
}

/**
 * Returns carousel element threshold.
 * @param {Object} target - target element.
 * @return {number} target threshold
 */
export function threshold(target) {
  const width = target.clientWidth;
  return width / 3;
}

/**
 * Carousel swipe event handler.
 * @param {Object} eventData data.
 * @param {number} length - slides length.
 * @param {number} dir - swipe direction.
 * @return {null | number} swipe length or null.
 */
export function swiped(eventData, length, dir = -1) {
  const t = threshold(eventData.event.target);
  const d = dir * eventData.deltaX;

  return d >= t ? length : null;
}

const basePosterUrl = (size) => `${setting.net.tmdbImageUrl}/t/p/w${size}/`;

export const posterUrl = (slide, size) => `${basePosterUrl(size)}${slide.poster_path}`;
