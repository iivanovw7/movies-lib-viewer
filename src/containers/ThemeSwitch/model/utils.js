/**
 * Module contains utils used in Theme Switch component
 * @module containers/ThemeSwitch/model/utils
 * @author Igor Ivanov
 */
import { lt, indexOf, inc, always, nth, ifElse, head, pipe, dec, __, length, keys, either } from 'ramda';

import { THEME_RESOURCES, DEFAULT_THEME } from './constants';

/**
 *  Gets string representations of keys out of theme object
 *  @return {Array.<string>} theme names list
 */
const themes = keys(THEME_RESOURCES);
/**
 * Gets index of current locale.
 * @type {f1}
 */
const themeIndex = indexOf(__, themes);
/**
 * Gets length of theme list.
 */
const themesLength = length(themes);
/**
 * Verifies if there is next theme available.
 * @type {f1}
 */
const hasNextTheme = lt(__, dec(themesLength));
/**
 * Returns firs theme, in case if no further themes available.
 */
const firstTheme = head(themes);

/**
 * Function selects next theme name.
 * Returns theme with next index if there is available one or first theme otherwise.
 * @type {Function|*}
 */
const newThemeKey = pipe(themeIndex, ifElse(hasNextTheme, pipe(inc, nth(__, themes)), always(firstTheme)));

/**
 * Returns either a new theme key or default key.
 * @type {Function|*}
 */
const nextThemeKey = either(newThemeKey, DEFAULT_THEME);

export default nextThemeKey;
