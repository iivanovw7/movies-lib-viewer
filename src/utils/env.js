/**
 * Module contains environment utils.
 * @module utils/env
 * @author Igor Ivanov
 */
/* eslint-disable import/prefer-default-export */
/**
 * Defines `IntersectionObserver` support.
 * @return {boolean}
 *  returns `true` if IntersectionObserver is fully supported and `false` otherwise.
 */
export const isObserverSupported = () =>
  'IntersectionObserver' in window &&
  'IntersectionObserverEntry' in window &&
  'intersectionRatio' in window.IntersectionObserverEntry.prototype;
/* eslint-enable import/prefer-default-export */
