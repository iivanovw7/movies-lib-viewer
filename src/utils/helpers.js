/**
 * Module contains helper functions.
 * @module utils/helpers
 */
import {
  equals,
  anyPass,
  isEmpty,
  isNil,
  __,
  path,
  ap,
  curry,
  useWith,
  split,
  map,
  of,
  is,
  find,
  propEq,
} from 'ramda';

/**
 * Checks if input value is `String`
 * @func isString
 * @return {Boolean}
 * @example
 * isString(String, 'foo') //=> true
 * isString(String, {a: 'foo'}) //=> false
 */
export const isString = is(String);

/**
 * Checks if input value is `Number`
 * @func isNumber
 * @return {Boolean}
 * @example
 * isString(Number, 123) //=> true
 * isString(Number, {a: 'foo'}) //=> false
 */
export const isNumber = is(Number);

/**
 * Checks if input value is `undefined`
 * @func isUndefined
 * @return {Boolean}
 * @example
 * isUndefined(1); //=> false
 * isUndefined(undefined); //=> true
 * isUndefined(null); //=> false
 */
export const isUndefined = equals(undefined);

/**
 * Returns `true` if the given value is its type's empty value, `null` or `undefined`
 * @func isNilOrEmpty
 * @return {Boolean}
 * @see {@link http://ramdajs.com/docs/#isEmpty|isEmpty}
 * @see {@link {@link http://ramdajs.com/docs/#isNil|isNil}}
 * @see {@link https://char0n.github.io/ramda-adjunct/2.23.0}
 * @example
 * isNilOrEmpty([1, 2, 3]); //=> false
 * isNilOrEmpty([]); //=> true
 * isNilOrEmpty(''); //=> true
 * isNilOrEmpty(null); //=> true
 * isNilOrEmpty(undefined): //=> true
 * isNilOrEmpty({}); //=> true
 * isNilOrEmpty({length: 0}); //=> false
 */
export const isNilOrEmpty = anyPass([isNil, isEmpty]);

/**
 * Acts as multiple path: arrays of paths in, array of values out. Preserves order.
 * @func paths
 * @param {Array} ps The property paths to fetch
 * @param {Object} obj The object to query
 * @return {Array} The corresponding values or partially applied function
 * @see {@link https://github.com/ramda/ramda/wiki/Cookbook#derivative-of-rprops-for-deep-fields}
 * @see {@link http://ramdajs.com/docs/#props|R.props}
 * @see {@link https://char0n.github.io/ramda-adjunct/2.23.0}
 * @example
 * const obj = {
 *   a: { b: { c: 1 } },
 *   x: 2,
 * };
 *
 * RA.paths([['a', 'b', 'c'], ['x']], obj); //=> [1, 2]
 */
export const paths = curry((ps, obj) => ap([path(__, obj)], ps));

/* eslint-disable react-hooks/rules-of-hooks */
/**
 * Parses dot paths, used in propsDotPath
 * @func dotPath
 * @memberOf {module:utils/helpers~propsDotPath}
 * @type {f1}
 */
const dotPath = useWith(path, [split('.')]);

/**
 * Derivative of R.props for deep fields
 * @func propsDotPath
 * @example
 * const obj = {
 *  a: { b: { c: 1 } },
 *  x: 2
 * };
 * propsDotPath(['a.b.c', 'x'], obj);
 * // => [ 1, 2 ]
 * @return {Array} array of props
 */
export const propsDotPath = useWith(ap, [map(dotPath), of]);

/**
 * Finds and returns first element that has `id` by default,
 *  or any other prop which is equal to passed in parameters.
 * @param {*} id - id value to be found
 * @param {*} [key=null] - specifies key to be found
 * @return {Function|*}
 *  should return found object in case there is a match
 */
export const findByPropValue = (id, key = 'id') => find(propEq(key, id));

/* eslint-enable react-hooks/rules-of-hooks */
