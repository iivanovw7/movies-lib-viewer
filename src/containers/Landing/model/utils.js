/**
 * Module contains landing page utils
 * @module containers/Landing/model/utils
 */
import { filter, pipe, prop } from 'ramda';

import { isString } from '../../../utils/helpers';

// eslint-disable-next-line import/prefer-default-export
export const withPosters = (list) => filter(pipe(prop('poster_path'), isString))(list);
