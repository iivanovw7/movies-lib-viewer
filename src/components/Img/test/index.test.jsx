/**
 * Module contains Img component test suite
 */
import { mount, shallow } from 'enzyme';
import React from 'react';
import { css } from 'styled-components';

import { noop } from '../../../__utils__/stub';
import { testName, mountWithTheme } from '../../../__utils__/utils';
import Img, { StyledImg } from '../index';

const baseProps = {
  src: 'link',
  alt: 'alt text',
  styling: [],
  className: 'test className',
  handleLoad: noop,
};

// prettier-ignore
describe(testName('Img', 'component test suite'), function ImgTestSuite() {
  // eslint-disable-next-line require-jsdoc
  const Composition = (props, shouldMount = false) => {
    // eslint-disable-next-line react/jsx-props-no-spreading
    const node = <Img {...props} />;

    return shouldMount ? mount(node) : shallow(node);
  };

  it('Should match snapshot', () => {
    const fakeProps = {
      ...baseProps,
    };
    const component = Composition(fakeProps);

    expect(component)
      .toMatchSnapshot();
  });

  it('Should render StyledImage', () => {
    const fakeProps = {
      ...baseProps,
    };
    const component = Composition(fakeProps);

    expect(component.find(StyledImg))
      .toHaveLength(1);
  });

  it('Should render component with props', () => {
    const StylesMock = css`
      height: 1.2em;
    `;
    const fakeProps = {
      ...baseProps,
      styling: StylesMock
    };
    const component = Composition(fakeProps);
    const image = component.find(StyledImg);

    expect(image.prop('src'))
      .toEqual(fakeProps.src);

    expect(image.prop('alt'))
      .toEqual(fakeProps.alt);

    expect(image.prop('className'))
      .toEqual(fakeProps.className);

    expect(image.prop('styling'))
      .toBe(StylesMock);
  });

  it('Should verify custom style rule on mount', () => {
    const Styles = css`
      height: 1.2em;
    `;
    const propsMock = {
      ...baseProps,
      styling: Styles
    };
    // eslint-disable-next-line react/jsx-props-no-spreading
    const component = mountWithTheme(<Img {...propsMock} />);

    expect(component.find(StyledImg))
      .toHaveStyleRule('height', '1.2em');
  });
});
