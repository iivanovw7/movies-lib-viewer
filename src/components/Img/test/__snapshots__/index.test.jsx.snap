// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[[33mImg[39m]: component test suite Should match snapshot 1`] = `
<Img__StyledImg
  alt="alt text"
  className="test className"
  onLoad={[Function]}
  src="link"
  styling={Array []}
/>
`;
